libraries/csharp-net22-runtime/context/Pipe/Module.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using Newtonsoft.Json;
+using System.Reflection;
+using Pipe;
+
+namespace Pipe
+{
+    public class Module
+    {
+
+        private string DllPath;
+        private Type PrimaryClassType;
+        private Type AlgoInputType;
+        private MethodInfo ApplyMethod;
+        
+        
+        public Module(Config config)
+        {
+            GetDllPath(config);
+            LoadClass(config);
+            GetApplyMethod();
+            GetAlgoInputType();
+        }
+        
+        
+        //TODO: Figure out if this is always the way to capture the dll file of a project. Are there other project formats?
+        private void GetDllPath(Config config)
+        {
+            string algoName = config.Algoname;
+            string algoPath = config.Algopath;
+            string boilerplate = "bin/Debug/netcoreapp2.2";
+            string fullpath = $"{algoPath}/{boilerplate}/{algoName}.dll";
+            Console.WriteLine(fullpath);
+            DllPath = fullpath;
+        }
+        private void LoadClass(Config config)
+        {
+            Assembly asm = Assembly.LoadFrom(DllPath);
+            PrimaryClassType = asm.GetType($"{config.Algoname}.{config.Algoname}");
+        }
+        private void GetApplyMethod()
+        {
+            // getting only public methods, as those are apply methods
+            MethodInfo applyMethod = PrimaryClassType.GetMethod("apply");
```

Comment:
In Java we allow more than 1 and try out multiple. Why are we limiting ourselves to 1? Reflection is hard and maybe the wrong choice here and would like to discuss that more to make sure we build something robust

Reply:
i think the java reflection stuff is actually super brittle and doesn't do the write thing in a couple of cases (it certainly used to be the case, i'm not sure if it still is).  i think part of the problem with java was that like booleans could get coerced into strings, and also that to pass multiple arguments into a java method we had to like call it with an array like `[1, "foo", false]` which feels a little less easy to understand than the json blob kind of inputs we usually give for python stuff `{"count":1, "name": "foo", "debug": false}`.  so yeah i think it's worth a discussion to figure out what approach would be the best.  i forget what rust does off the top of my head - @anowell ?

Reply:
We're limiting to 1 for simplicity of a PoC, we can have any number, and we can be way more explicit around type checking. We can even have apply's in different modules, but we have to set some kind of structure somewhere. For a basic, fully functional PoC (You can do anything you want with just 1 apply, etc), I think this is ok as is.

WRT the type coercsion issue.. it's still problematic, I'm not sure yet what a solution looks like. I would agree that having a deeper discussion makes sense here, although again outside of scope for a PoC.

Reply:
Refactored into a static compilation system in latest commit.